cmd/kubeadm/app/util/etcd/etcd_test.go

```diff
@@ -494,3 +495,221 @@ func TestClient_GetMemberID(t *testing.T) {
 		})
 	}
 }
+
+func TestListMembers(t *testing.T) {
+	type fields struct {
+		Endpoints     []string
+		newEtcdClient func(endpoints []string) (etcdClient, error)
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		want   []Member
+	}{
+		{
+			name: "PeerURLs are empty",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{}
+					return f, nil
+				},
+			},
+			want: []Member{},
+		},
+		{
+			name: "PeerURLs are exist",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								ID:   1,
+								Name: "member1",
+								PeerURLs: []string{
+									"https://member1:2380",
+								},
+							},
+							{
+								ID:   2,
+								Name: "member2",
+								PeerURLs: []string{
+									"https://member2:2380",
+								},
+							},
+						},
+					}
+					return f, nil
+				},
+			},
+			want: []Member{
+				{
+					Name:    "member1",
+					PeerURL: "https://member1:2380",
+				},
+				{
+					Name:    "member2",
+					PeerURL: "https://member2:2380",
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := &Client{
+				Endpoints:     tt.fields.Endpoints,
+				newEtcdClient: tt.fields.newEtcdClient,
+			}
+			got, err := c.ListMembers()
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("ListMembers() = %v, want %v", got, tt.want)
+			}
+			assert.NoError(t, err)
```

Comment:
is it possible to have a test case that causes an error?
(i.e. we can have an wantError bool and check that as well)

Reply:
The function tested in this PR may return error only here, but since it will be retried after returning an error here, in order to avoid affecting other places, I did not test the situation where error is returned. Could  you give any advices? Thanks!
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/etcd/etcd.go#L281

Reply:
an error can be returned in the lines of "expect client error from endpoints":
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/etcd/etcd.go#L282

but there is another problem in this unit test.
ideally the function:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/etcd/etcd.go#L277

should accept a custom backoff during unit testing and not use etcdBackoff:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/etcd/etcd.go#L281C33-L281C44

it could be:
```
func (c *Client) listMembers(backoff *wait.Backoff) (*clientv3.MemberListResponse, error) {
   if backoff == nil {
      backoff = etcdBackoff
   }
```

Reply:
Thanks, do I need to modify `etcd.go` code in this PR?

Reply:
yes, maybe this approach is better
https://github.com/kubernetes/kubernetes/pull/122356#discussion_r1430373179